docs: improve phrasing around package dist path updates (#7285)
typescript-eslint/typescript-eslint: packages/website/blog/2023-07-09-announcing-typescript-eslint-v6.md

### Package Exports

> The v5 `@typescript-eslint/` packages don't use [Node.js package.json exports](https://nodejs.org/api/packages.html#package-entry-points), which allows someone to import any file in any package by directly referencing a path within the `dist` directory.

Reviewer: 🤔 I don't see any advantage to the new sentence form here. It's a bit longer and conveys the same information. Is there a reason you'd prefer the new form?

Author: 1) the word "directory" is more agnostic and should be generally preferred over "folder", since "folder" is specifically a Windows word.
2) `dist` should probably be in code ticks to signify that it is a directory instead of a word.
3) "anyone --> any file --> any package" might be a bit confusing.
the important part is the "any file in any package" part i think.
we are not just talking about "anyone", we are specifically talking about "end-users who invoke the TSESLint API directly in their code", NOT people who just use the TSESLint plugin more conventionally.
maybe this can be workshopped further.

Reviewer: Got it - you raise good points!
* I think the difference between "directory" and "folder" is pretty low priority for us, and "directory" is longer to read ever so slightly... but yeah makes sense
* +1 on the ticks
* Agreed that the original phrasing is confusing and also that I do want to simplify the new phrasing a bit...

Maybe:

```suggestion
The v5 `@typescript-eslint/*` packages don't use [Node.js package.json exports](https://nodejs.org/api/packages.html#package-entry-points), which allows importing any file in any package by directly referencing a path within the package's `dist/` directory.
```

How do you feel about that?

Author: lgtm

That presents a few issues for developers:

- It can be unclear which of many potential import paths to use.
- TypeScript sometimes suggests importing types or values meant to be private.
- Consumers using deep import paths might get broken code after internal refactors rename files.

As of [feat: add package.json exports for public packages](https://github.com/typescript-eslint/typescript-eslint/pull/6458), `@typescript-eslint/*` packages now use `exports` to prevent importing internal file paths.
Developers must now mostly import directly from the package names. For example, to fix the code from the previous example, you would use the following:

```ts
import * as TSESLint from '@typescript-eslint/utils';
```

See [RFC: Use package.json exports to "hide" the dist folder for packages and control our exported surface-area](https://github.com/typescript-eslint/typescript-eslint/discussions/6015) for more backing context.

> Note that if you update your imports and you still get an error from TypeScript saying "Cannot find module '@typescript-eslint/utils' or its corresponding type declarations", then you might need to change the value of `moduleResolution` to `node16` in your TypeScript config, as detailed in [this issue](https://github.com/typescript-eslint/typescript-eslint/issues/7279).

Reviewer: > to `node16`

I've learned the hard way not to be prescriptive about project-dependent compiler options 😄. Some folks would want `bundler` or `nodenext`.

How does this look to you?

> Note that if you update your imports and you still get an error from TypeScript saying "Cannot find module '@typescript-eslint/utils' or its corresponding type declarations", then you might need to change the value of `moduleResolution` to `node16` in your TypeScript config, as detailed in [this issue](https://github.com/typescript-eslint/typescript-eslint/issues/7279).

Reviewer: > to `node16`

I've learned the hard way not to be prescriptive about project-dependent compiler options 😄. Some folks would want `bundler` or `nodenext`.

How does this look to you?

```suggestion
:::note
If you update your imports and you still get an error from TypeScript saying _`"Cannot find module '@typescript-eslint/...' or its corresponding type declarations"`_, then you might need to change the value of `moduleResolution` in your TypeScript config. See [this tracking issue for `package.json` exports types](https://github.com/typescript-eslint/typescript-eslint/issues/7284).
:::
```

Also lumping in a few changes in that suggestion:

* Switching to a [Docusaurus `node` block](https://docusaurus.io/docs/markdown-features/admonitions), because they're great
* Using a [more descriptive link name for accessibility](https://www.chhs.colostate.edu/accessibility/best-practices-how-tos/descriptive-links)
* Using a new pinned issue I filed just now (so folks don't have to read through longer discussions)

Author: lgtm!

That presents a few issues for developers:

- It can be unclear which of many potential import paths to use.
- TypeScript sometimes suggests importing types or values meant to be private.

> - Consumers using deep import paths might get broken code after internal refactors rename files.

Reviewer: Speaking directly to this one: I still think the original phrasing is more clear:

* The "can" and "might" mean about the same
* It's not the code that's broken, it's the runtime and import paths

Author: ok, i reverted the change

As of [feat: add package.json exports for public packages](https://github.com/typescript-eslint/typescript-eslint/pull/6458), `@typescript-eslint/*` packages now use `exports` to prevent importing internal file paths.

> Developers must now import directly from the package names. e.g.:

Reviewer: I believe this isn't _always_ the case - are still some deeper paths.

```suggestion
Developers must now mostly import directly from the package names, e.g.:
```

> Developers must now mostly import directly from the package names. For example, to fix the code from the previous example, you would use the following:

Reviewer: > you

I think the `.e.g.` (or, if we don't like the abbreviation, a `for example`) says just about the same as this new sentence. Let's revert this change.

As an aside: for these general-purpose docs I don't think we want to phrase them as a personal journey. I in the past received feedback in -I think- the ESLint core repo that _"you"_ can come across as a bit pointed, even accusatory. I agree with that point of view.

Author: ok, i reverted it

```ts
import * as TSESLint from '@typescript-eslint/utils';
```

See [RFC: Use package.json exports to "hide" the dist folder for packages and control our exported surface-area](https://github.com/typescript-eslint/typescript-eslint/discussions/6015) for more backing context.

:::note
If you update your imports and you still get an error from TypeScript saying _`"Cannot find module '@typescript-eslint/...' or its corresponding type declarations"`_, then you might need to change the value of `moduleResolution` in your TypeScript config. See [this tracking issue for `package.json` exports types](https://github.com/typescript-eslint/typescript-eslint/issues/7284).

> :::

Reviewer: [Praise] YES +1

That presents a few issues for developers:

- It can be unclear which of many potential import paths to use.
- TypeScript sometimes suggests importing types or values meant to be private.
- Consumers using deep import paths can be broken by internal refactors that rename files.

As of [feat: add package.json exports for public packages](https://github.com/typescript-eslint/typescript-eslint/pull/6458), `@typescript-eslint/*` packages now use `exports` to prevent importing internal file paths.
Developers must now mostly import directly from the package names, e.g.:

```ts

> import * as TSESLint from '@typescript-eslint/utils';

Reviewer: Checking with @bradzacher as previous editor of this line - what is the right import we'd to convey here?

Reviewer 2: This has already been fixed in main:
https://github.com/typescript-eslint/typescript-eslint/blob/6addee9b4ef668cd2e026130083a3a783b327e9e/packages/website/blog/2023-07-09-announcing-typescript-eslint-v6.md#L636-L641

Author: Shouldn't we recommend this instead?

```ts
import { TSESLint } from '@typescript-eslint/utils'; 
```

IMO star imports are not great.

Reviewer 2: It depends - you'll find a lot of people we're actually doing things like this before:

```ts
import { RuleModule } from '@typescript-eslint/utils/dist/ts-eslint' ;
```

So for those usecases switching to

```ts
import { RuleModule } from '@typescript-eslint/utils/ts-eslint' ;
```

Is the better migration!

However if someone was using the `TSESLint` namespace before then it's unlikely that they were deeply importing it - so they wouldn't need to update anything.

If we were to do anything to update this here - we'd probably want to update the examples to be "realer".

Author: > we'd probably want to update the examples to be "realer".

Ok, I updated the PR to say this:

```ts
// import * as TSESLint from '@typescript-eslint/utils/dist/ts-eslint';
// -->
import { TSESLint } from '@typescript-eslint/utils';

// import { RuleModule } from '@typescript-eslint/utils/dist/ts-eslint';
// -->
import { RuleModule } from '@typescript-eslint/utils/ts-eslint';

// import { AST_NODE_TYPES } from "@typescript-eslint/types/dist/generated/ast-spec";
// -->
import { AST_NODE_TYPES } from "@typescript-eslint/types";
```

Reviewer: cc @bradzacher?
